urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/LzmaBench.java

```diff
@@ -0,0 +1,480 @@
+            System.out.print(" ");
+        }
+        System.out.print(s);
+    }
+    
+    static void PrintRating(long rating)
+    {
+        PrintValue(rating / 1000000);
+        System.out.print(" MIPS");
+    }
+    
+    static void PrintResults(int dictionarySize, long elapsedTime, long size,
+            boolean decompressMode, long secondSize)
+    {
+        final long speed = MyMultDiv64(size, elapsedTime);
+        PrintValue(speed / 1024);
+        System.out.print(" KB/s  ");
+        long rating;
+        if (decompressMode)
+        {
+            rating = GetDecompressRating(elapsedTime, size, secondSize);
+        }
+        else
+        {
+            rating = GetCompressRating(dictionarySize, elapsedTime, size);
+        }
+        PrintRating(rating);
+    }
+    
+    static public int LzmaBenchmark(int numIterations, int dictionarySize)
+            throws Exception
+    {
+        if (numIterations <= 0)
+        {
+            return 0;
+        }
+        if (dictionarySize < 1 << 18)
+        {
+            System.out
+                    .println("\nError: dictionary size for benchmark must be >= 18 (256 KB)");
+            return 1;
+        }
+        System.out
+                .print("\n       Compressing                Decompressing\n\n");
+        
+        final SevenZip.Compression.LZMA.Encoder encoder = new SevenZip.Compression.LZMA.Encoder();
+        final SevenZip.Compression.LZMA.Decoder decoder = new SevenZip.Compression.LZMA.Decoder();
+        
+        if (!encoder.SetDictionarySize(dictionarySize))
+        {
+            throw new Exception("Incorrect dictionary size");
+        }
+        
+        final int kBufferSize = dictionarySize + kAdditionalSize;
+        final int kCompressedBufferSize = kBufferSize / 2
+                + kCompressedAdditionalSize;
+        
+        final ByteArrayOutputStream propStream = new ByteArrayOutputStream();
+        encoder.WriteCoderProperties(propStream);
+        final byte[] propArray = propStream.toByteArray();
+        decoder.SetDecoderProperties(propArray);
+        
+        final CBenchRandomGenerator rg = new CBenchRandomGenerator();
+        
+        rg.Set(kBufferSize);
+        rg.Generate();
+        final CRC crc = new CRC();
+        crc.Init();
+        crc.Update(rg.Buffer, 0, rg.BufferSize);
+        
+        final CProgressInfo progressInfo = new CProgressInfo();
+        progressInfo.ApprovedStart = dictionarySize;
+        
+        long totalBenchSize = 0;
+        long totalEncodeTime = 0;
+        long totalDecodeTime = 0;
+        long totalCompressedSize = 0;
+        
+        final MyInputStream inStream = new MyInputStream(rg.Buffer,
+                rg.BufferSize);
+        
+        final byte[] compressedBuffer = new byte[kCompressedBufferSize];
+        final MyOutputStream compressedStream = new MyOutputStream(
+                compressedBuffer);
+        final CrcOutStream crcOutStream = new CrcOutStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *crcOutStream*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).